File: csharp/squeaky-clean/SqueakyClean.cs
using System.Text;

public static class Identifier
{
    public static string Clean(string identifier)
    {
        if (string.IsNullOrWhiteSpace(identifier)) return "";
        var sb = new StringBuilder();
        var chars = identifier.ToCharArray();
        (char, char) greek = ('α', 'ω');

        var toCamelCase = false;

        foreach (var (ch, index) in chars.Select((ch, index) => (ch, index)))
        {
            if (toCamelCase)
            {
                sb.Append(Char.ToUpper(ch));
                toCamelCase = false;
                continue;
            }
            if (Char.IsWhiteSpace(ch))
            {
                sb.Append('_');
                continue;
            }
            if (Char.IsControl(ch))
            {
                sb.Append("CTRL");
                continue;
            }
            if (ch == '-' && index < chars.Length - 1 && Char.IsLetter(chars[index + 1]))
            {
                toCamelCase = true;
                continue;
            }
            if (ch >= greek.Item1 && ch <= greek.Item2) continue;
            if (!Char.IsLetter(ch)) continue;
            sb.Append(ch);
        }

        return sb.ToString();
    }
}

File: java/squeaky-clean/src/main/java/SqueakyClean.java
import java.lang.Character;

class SqueakyClean {
    static String clean(String identifier) {
        char[] chs = identifier.toCharArray();
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < chs.length; i++) {
            switch (chs[i]) {
                case ' ':
                    sb.append('_');
                    break;
                case '-':
                    if (i + 1 < chs.length && Character.isLetter(chs[i + 1])) {
                        sb.append(Character.toUpperCase(chs[i + 1]));
                        i += 1;
                    }
                    break;
                case '4':
                    sb.append('a');
                    break;
                case '3':
                    sb.append('e');
                    break;
                case '0':
                    sb.append('o');
                    break;
                case '1':
                    sb.append('l');
                    break;
                case '7':
                    sb.append('t');
                    break;
                default:
                    if (Character.isLetter(chs[i]))
                        sb.append(chs[i]);
            }
        }
        return sb.toString();
    }
}
